addOns/pscanrules/src/main/java/org/zaproxy/zap/extension/pscanrules/SameOriginMethodExecutionScanRule.java

```diff
@@ -0,0 +1,227 @@
+                    createAlert(
+                                    "jsonp",
+                                    (sfdHeader != null && sfdHeader.equals("script"))
+                                            ? Alert.CONFIDENCE_HIGH
+                                            : Alert.CONFIDENCE_MEDIUM,
+                                    param.getValue(),
+                                    refererHeader)
+                            .raise();
+                }
+            }
+        }
+    }
+
+    private AlertBuilder createAlert(
+            String type, int confidence, String evidence, String otherInfoParam) {
+        return newAlert()
+                .setRisk(Alert.RISK_HIGH)
+                .setConfidence(confidence)
+                .setDescription(Constant.messages.getString(MESSAGE_PREFIX + "desc"))
+                .setOtherInfo(
+                        Constant.messages.getString(
+                                MESSAGE_PREFIX + type + ".extrainfo", otherInfoParam))
+                .setEvidence(evidence)
+                .setSolution(Constant.messages.getString(MESSAGE_PREFIX + "soln"))
+                .setCweId(359) // CWE-359: Exposure of Private Information ('Privacy Violation')
+                .setWascId(13); // WASC-13: Information Leakage
+    }
+
+    private void performHeuristicsCheck(
+            HtmlParameter param, List<Element> scriptTags, boolean isSwf) {
+
+        LOGGER.info(param.toString());
+        if (CALLBACK_PARAMS.contains(param.getName()) && isSwf) {
+            createAlert("swf", Alert.CONFIDENCE_MEDIUM, param.getName(), param.getName());
+        }
+
+        for (Element tag : scriptTags) {
+            String src = tag.getAttributeValue("src");
+            LOGGER.info(src);
+            for (String callbackParam : CALLBACK_PARAMS) {
+                String expectedUrlParam = callbackParam + "=" + param.getValue();
+                LOGGER.info(expectedUrlParam);
+
+                if (src.contains(expectedUrlParam)) {
+                    createAlert(
+                                    "heuristics",
+                                    Alert.CONFIDENCE_MEDIUM,
+                                    expectedUrlParam,
+                                    param.getName())
+                            .raise();
+                }
+            }
+        }
+    }
+
+    private TreeSet<HtmlParameter> getUrlParameters(HttpMessage newReq) {
+        try {
+            newReq.getRequestHeader()
+                    .setURI(
+                            new URI(
+                                    URLDecoder.decode(
+                                            newReq.getRequestHeader().getURI().toString(),
+                                            StandardCharsets.UTF_8.toString()),
+                                    true));
+        } catch (URIException | UnsupportedEncodingException e) {
+            e.printStackTrace();
```

Comment:
This should use LOGGER, if really necessary.

Reply:
Done :white_check_mark: